### ISSUE TITLE
Possibly wrong default identity value for ``maximum<T>`` reduction

### ISSUE THREAD

**Issue by stukowski** [AdaptiveCpp/AdaptiveCpp#1310]
**Bug summary**

I came across [this code](https://github.com/AdaptiveCpp/AdaptiveCpp/blob/7041cdf31521571d1cc7aad4de6fe7dc27593317/include/hipSYCL/algorithms/numeric.hpp#L67C1-L67C90) in AdaptiveCpp, which makes me think that a wrong default identity value may be used for the ``sycl::maximum`` reduction:

```c++
HIPSYCL_ALGORITHMS_DEFINE_KNOWN_IDENTITY(sycl::minimum<T>, std::numeric_limits<T>::max())
HIPSYCL_ALGORITHMS_DEFINE_KNOWN_IDENTITY(sycl::maximum<T>, std::numeric_limits<T>::min())
``` 

Shouldn't ``numeric_limits<T>::lowest()`` be used instead of ``min()``? Note that ``min()`` does *not* yield the lowest value of type ``T``.

So far this is only a theoretical observation, but I would expect that this will likely lead to incorrect results in practice when performing a maximum reduction over a set of floating-point numbers that are all negative. Let me know if you want me to test my hypothesis.

**Comment by illuhad** [comment]
I agree that there might be an issue here. I think floats require a more complicated handling in general. For example, we should also take into account what happens if `inf` values etc are involved.

Sure, go ahead and validate / create a PR :)

**Comment by stukowski** [comment]
Yes, good point, for floating-point types `inf` values should be used as identity values.

Here is a link to the logic found in DPC++, which shows that, for the `sycl::maximum` reduction, either `-infinity()` or `lowest()` is used depending on the numeric type:

https://github.com/intel/llvm/blob/c5acdff7d681133f874250be5562106652edfa80/sycl/include/sycl/known_identity.hpp#L299C1-L309C3

**Comment by illuhad** [comment]
Does that still work if `-ffast-math` is used during compilation?

**Comment by stukowski** [comment]
I am not an expert on this topic, but a quick experiment with Godbolt shows that the `-ffast-math` option (implying `-ffinite-math-only`) does not break the use of `numeric_limits::infinity()` as identity value. But some compilers may emit a compiler warning. So it could make sense to disable the use of `numeric_limits::infinity()` if the `__FINITE_MATH_ONLY__` macro is defined.

https://godbolt.org/z/8cr5j5jbq
